// sprotty-theia/lib/sprotty/languageserver/workspace-edit-command.js
"use strict";
/********************************************************************************
 * Copyright (c) 2018 TypeFox and others.
 *
 * This program and the accompanying materials are made available under the
 * terms of the Eclipse Public License v. 2.0 which is available at
 * http://www.eclipse.org/legal/epl-2.0.
 *
 * This Source Code may also be made available under the following Secondary
 * Licenses when the conditions for such availability set forth in the Eclipse
 * Public License v. 2.0 are satisfied: GNU General Public License, version 2
 * with the GNU Classpath Exception which is available at
 * https://www.gnu.org/software/classpath/license.html.
 *
 * SPDX-License-Identifier: EPL-2.0 OR GPL-2.0 WITH Classpath-exception-2.0
 ********************************************************************************/
var __extends = (this && this.__extends) || (function () {
    var extendStatics = Object.setPrototypeOf ||
        ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
        function (d, b) { for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p]; };
    return function (d, b) {
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
var __metadata = (this && this.__metadata) || function (k, v) {
    if (typeof Reflect === "object" && typeof Reflect.metadata === "function") return Reflect.metadata(k, v);
};
var __param = (this && this.__param) || function (paramIndex, decorator) {
    return function (target, key) { decorator(target, key, paramIndex); }
};
Object.defineProperty(exports, "__esModule", { value: true });
var inversify_1 = require("inversify");
var sprotty_1 = require("sprotty");
var ls_theia_diagram_server_1 = require("./ls-theia-diagram-server");
var AbstractWorkspaceEditCommand = /** @class */ (function (_super) {
    __extends(AbstractWorkspaceEditCommand, _super);
    function AbstractWorkspaceEditCommand() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    Object.defineProperty(AbstractWorkspaceEditCommand.prototype, "workspace", {
        get: function () {
            return this.diagramServer.connector.workspace;
        },
        enumerable: true,
        configurable: true
    });
    AbstractWorkspaceEditCommand.prototype.execute = function (context) {
        this.workspaceEdit = this.createWorkspaceEdit(context);
        this.workspace.applyEdit(this.workspaceEdit, { mode: 'open' });
        return context.root;
    };
    AbstractWorkspaceEditCommand.prototype.undo = function (context) {
        // TODO implement revert workspace edit
        return context.root;
    };
    AbstractWorkspaceEditCommand.prototype.redo = function (context) {
        // TODO implement revert workspace edit
        return context.root;
    };
    __decorate([
        inversify_1.inject(ls_theia_diagram_server_1.LSTheiaDiagramServer),
        __metadata("design:type", ls_theia_diagram_server_1.LSTheiaDiagramServer)
    ], AbstractWorkspaceEditCommand.prototype, "diagramServer", void 0);
    AbstractWorkspaceEditCommand = __decorate([
        inversify_1.injectable()
    ], AbstractWorkspaceEditCommand);
    return AbstractWorkspaceEditCommand;
}(sprotty_1.Command));
exports.AbstractWorkspaceEditCommand = AbstractWorkspaceEditCommand;
/**
 * This is a client only action, so it does not have to be serializable
 */
var WorkspaceEditAction = /** @class */ (function () {
    function WorkspaceEditAction(workspaceEdit) {
        this.workspaceEdit = workspaceEdit;
        this.kind = WorkspaceEditAction.KIND;
    }
    WorkspaceEditAction.KIND = 'workspaceEdit';
    return WorkspaceEditAction;
}());
exports.WorkspaceEditAction = WorkspaceEditAction;
var WorkspaceEditCommand = /** @class */ (function (_super) {
    __extends(WorkspaceEditCommand, _super);
    function WorkspaceEditCommand(action) {
        var _this = _super.call(this) || this;
        _this.action = action;
        return _this;
    }
    WorkspaceEditCommand.prototype.createWorkspaceEdit = function (context) {
        return this.action.workspaceEdit;
    };
    WorkspaceEditCommand.KIND = WorkspaceEditAction.KIND;
    WorkspaceEditCommand = __decorate([
        inversify_1.injectable(),
        __param(0, inversify_1.inject(sprotty_1.TYPES.Action)),
        __metadata("design:paramtypes", [WorkspaceEditAction])
    ], WorkspaceEditCommand);
    return WorkspaceEditCommand;
}(AbstractWorkspaceEditCommand));
exports.WorkspaceEditCommand = WorkspaceEditCommand;
