// @theia/php/lib/node/php-contribution.js
"use strict";
/*
 * Copyright (C) 2017 TypeFox and others.
 *
 * Licensed under the Apache License, Version 2.0 (the "License"); you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at http://www.apache.org/licenses/LICENSE-2.0
 */
var __extends = (this && this.__extends) || (function () {
    var extendStatics = function (d, b) {
        extendStatics = Object.setPrototypeOf ||
            ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
            function (d, b) { for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p]; };
        return extendStatics(d, b);
    };
    return function (d, b) {
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : new P(function (resolve) { resolve(result.value); }).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __generator = (this && this.__generator) || function (thisArg, body) {
    var _ = { label: 0, sent: function() { if (t[0] & 1) throw t[1]; return t[1]; }, trys: [], ops: [] }, f, y, t, g;
    return g = { next: verb(0), "throw": verb(1), "return": verb(2) }, typeof Symbol === "function" && (g[Symbol.iterator] = function() { return this; }), g;
    function verb(n) { return function (v) { return step([n, v]); }; }
    function step(op) {
        if (f) throw new TypeError("Generator is already executing.");
        while (_) try {
            if (f = 1, y && (t = op[0] & 2 ? y["return"] : op[0] ? y["throw"] || ((t = y["return"]) && t.call(y), 0) : y.next) && !(t = t.call(y, op[1])).done) return t;
            if (y = 0, t) op = [op[0] & 2, t.value];
            switch (op[0]) {
                case 0: case 1: t = op; break;
                case 4: _.label++; return { value: op[1], done: false };
                case 5: _.label++; y = op[1]; op = [0]; continue;
                case 7: op = _.ops.pop(); _.trys.pop(); continue;
                default:
                    if (!(t = _.trys, t = t.length > 0 && t[t.length - 1]) && (op[0] === 6 || op[0] === 2)) { _ = 0; continue; }
                    if (op[0] === 3 && (!t || (op[1] > t[0] && op[1] < t[3]))) { _.label = op[1]; break; }
                    if (op[0] === 6 && _.label < t[1]) { _.label = t[1]; t = op; break; }
                    if (t && _.label < t[2]) { _.label = t[2]; _.ops.push(op); break; }
                    if (t[2]) _.ops.pop();
                    _.trys.pop(); continue;
            }
            op = body.call(thisArg, _);
        } catch (e) { op = [6, e]; y = 0; } finally { f = t = 0; }
        if (op[0] & 5) throw op[1]; return { value: op[0] ? op[1] : void 0, done: true };
    }
};
Object.defineProperty(exports, "__esModule", { value: true });
//import * as os from 'os';
var path = require("path");
// import * as glob from 'glob';
var inversify_1 = require("inversify");
// import { DEBUG_MODE } from '@theia/core/lib/node';
var node_1 = require("@theia/languages/lib/node");
var common_1 = require("../common");
var utils_1 = require("@theia/process/lib/node/utils");
// export type ConfigurationType = 'config_win' | 'config_mac' | 'config_linux';
// export const configurations = new Map<typeof process.platform, ConfigurationType>();
// configurations.set('darwin', 'config_mac');
// configurations.set('win32', 'config_win');
// configurations.set('linux', 'config_linux');
var PHPContribution = /** @class */ (function (_super) {
    __extends(PHPContribution, _super);
    function PHPContribution() {
        var _this = _super !== null && _super.apply(this, arguments) || this;
        _this.id = common_1.PHP_LANGUAGE_ID;
        _this.name = common_1.PHP_LANGUAGE_NAME;
        return _this;
    }
    PHPContribution.prototype.start = function (clientConnection) {
        return __awaiter(this, void 0, void 0, function () {
            var lsp, command, args, phpLsCommand, serverConnection;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0:
                        lsp = path.join(__dirname, '../..', 'vendor', 'felixfbecker', 'language-server', 'bin', 'php-language-server.php');
                        command = 'php';
                        args = [
                            lsp
                        ];
                        phpLsCommand = process.env.PHP_LS_COMMAND;
                        if (phpLsCommand) {
                            command = phpLsCommand;
                            args = utils_1.parseArgs(process.env.PHP_LS_ARGS || '');
                        }
                        console.info("starting PHP language server :)");
                        console.info(lsp);
                        return [4 /*yield*/, this.createProcessStreamConnectionAsync(command, args, this.getSpawnOptions())];
                    case 1:
                        serverConnection = _a.sent();
                        // serverConnection.reader.onError(err => console.log(err));
                        this.forward(clientConnection, serverConnection);
                        return [2 /*return*/];
                }
            });
        });
    };
    PHPContribution.prototype.getSpawnOptions = function () {
        return undefined;
    };
    PHPContribution.prototype.onDidFailSpawnProcess = function (error) {
        _super.prototype.onDidFailSpawnProcess.call(this, error);
        console.error("Error starting php language server.");
        console.error("Please make sure it is installed on your system.");
    };
    PHPContribution = __decorate([
        inversify_1.injectable()
    ], PHPContribution);
    return PHPContribution;
}(node_1.BaseLanguageServerContribution));
exports.PHPContribution = PHPContribution;
