// yang-sprotty/lib/yang-models.js
"use strict";
/*
 * Copyright (C) 2017 TypeFox and others.
 *
 * Licensed under the Apache License, Version 2.0 (the "License"); you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at http://www.apache.org/licenses/LICENSE-2.0
 */
var __extends = (this && this.__extends) || (function () {
    var extendStatics = Object.setPrototypeOf ||
        ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
        function (d, b) { for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p]; };
    return function (d, b) {
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
var sprotty_1 = require("sprotty");
var YangNode = /** @class */ (function (_super) {
    __extends(YangNode, _super);
    function YangNode() {
        var _this = _super !== null && _super.apply(this, arguments) || this;
        _this.strokeWidth = 1;
        return _this;
    }
    YangNode.prototype.hasFeature = function (feature) {
        return feature === sprotty_1.selectFeature || feature === sprotty_1.boundsFeature
            || feature === sprotty_1.layoutContainerFeature || feature === sprotty_1.fadeFeature || feature === sprotty_1.hoverFeedbackFeature
            || feature === sprotty_1.popupFeature || (feature === sprotty_1.openFeature && this.trace !== undefined);
    };
    return YangNode;
}(sprotty_1.RectangularNode));
exports.YangNode = YangNode;
var ModuleNode = /** @class */ (function (_super) {
    __extends(ModuleNode, _super);
    function ModuleNode() {
        var _this = _super !== null && _super.apply(this, arguments) || this;
        _this.expanded = false;
        return _this;
    }
    ModuleNode.prototype.hasFeature = function (feature) {
        return feature === sprotty_1.expandFeature || _super.prototype.hasFeature.call(this, feature);
    };
    return ModuleNode;
}(YangNode));
exports.ModuleNode = ModuleNode;
var YangHeaderNode = /** @class */ (function (_super) {
    __extends(YangHeaderNode, _super);
    function YangHeaderNode() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    return YangHeaderNode;
}(sprotty_1.SCompartment));
exports.YangHeaderNode = YangHeaderNode;
var YangLabel = /** @class */ (function (_super) {
    __extends(YangLabel, _super);
    function YangLabel() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    YangLabel.prototype.hasFeature = function (feature) {
        return _super.prototype.hasFeature.call(this, feature) || feature === sprotty_1.selectFeature || (feature === sprotty_1.openFeature && this.trace !== undefined);
    };
    return YangLabel;
}(sprotty_1.SLabel));
exports.YangLabel = YangLabel;
var Tag = /** @class */ (function (_super) {
    __extends(Tag, _super);
    function Tag() {
        var _this = _super !== null && _super.apply(this, arguments) || this;
        _this.size = {
            width: 24,
            height: 24
        };
        return _this;
    }
    Tag.prototype.hasFeature = function (feature) {
        return feature === sprotty_1.boundsFeature || feature === sprotty_1.layoutContainerFeature || feature === sprotty_1.layoutableChildFeature || feature === sprotty_1.fadeFeature;
    };
    return Tag;
}(sprotty_1.SShapeElement));
exports.Tag = Tag;
