// @theia/messages/lib/browser/notifications-contribution.js
"use strict";
/********************************************************************************
 * Copyright (C) 2019 TypeFox and others.
 *
 * This program and the accompanying materials are made available under the
 * terms of the Eclipse Public License v. 2.0 which is available at
 * http://www.eclipse.org/legal/epl-2.0.
 *
 * This Source Code may also be made available under the following Secondary
 * Licenses when the conditions for such availability set forth in the Eclipse
 * Public License v. 2.0 are satisfied: GNU General Public License, version 2
 * with the GNU Classpath Exception which is available at
 * https://www.gnu.org/software/classpath/license.html.
 *
 * SPDX-License-Identifier: EPL-2.0 OR GPL-2.0 WITH Classpath-exception-2.0
 ********************************************************************************/
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
var __metadata = (this && this.__metadata) || function (k, v) {
    if (typeof Reflect === "object" && typeof Reflect.metadata === "function") return Reflect.metadata(k, v);
};
Object.defineProperty(exports, "__esModule", { value: true });
var inversify_1 = require("inversify");
var browser_1 = require("@theia/core/lib/browser");
var notifications_commands_1 = require("./notifications-commands");
var notifications_manager_1 = require("./notifications-manager");
var notifications_renderer_1 = require("./notifications-renderer");
var NotificationsContribution = /** @class */ (function () {
    function NotificationsContribution() {
        this.id = 'theia-notification-center';
    }
    NotificationsContribution.prototype.onStart = function (_app) {
        this.createStatusBarItem();
    };
    NotificationsContribution.prototype.createStatusBarItem = function () {
        var _this = this;
        this.updateStatusBarItem();
        this.manager.onUpdated(function (e) { return _this.updateStatusBarItem(e.notifications.length); });
    };
    NotificationsContribution.prototype.updateStatusBarItem = function (count) {
        if (count === void 0) { count = 0; }
        this.statusBar.setElement(this.id, {
            text: this.getStatusBarItemText(count),
            alignment: browser_1.StatusBarAlignment.RIGHT,
            priority: -900,
            command: notifications_commands_1.NotificationsCommands.TOGGLE.id
        });
    };
    NotificationsContribution.prototype.getStatusBarItemText = function (count) {
        return "$(bell) " + (count ? " " + count : '');
    };
    NotificationsContribution.prototype.registerCommands = function (commands) {
        var _this = this;
        commands.registerCommand(notifications_commands_1.NotificationsCommands.TOGGLE, {
            isEnabled: function () { return true; },
            execute: function () { return _this.manager.toggleCenter(); }
        });
        commands.registerCommand(notifications_commands_1.NotificationsCommands.HIDE, {
            execute: function () { return _this.manager.hide(); }
        });
        commands.registerCommand(notifications_commands_1.NotificationsCommands.CLEAR_ALL, {
            execute: function () { return _this.manager.clearAll(); }
        });
    };
    NotificationsContribution.prototype.registerKeybindings = function (keybindings) {
        keybindings.registerKeybinding({
            command: notifications_commands_1.NotificationsCommands.HIDE.id,
            context: NotificationsKeybindingContext.notificationsVisible,
            keybinding: 'esc'
        });
    };
    __decorate([
        inversify_1.inject(notifications_manager_1.NotificationManager),
        __metadata("design:type", notifications_manager_1.NotificationManager)
    ], NotificationsContribution.prototype, "manager", void 0);
    __decorate([
        inversify_1.inject(notifications_renderer_1.NotificationsRenderer),
        __metadata("design:type", notifications_renderer_1.NotificationsRenderer)
    ], NotificationsContribution.prototype, "notificationsRenderer", void 0);
    __decorate([
        inversify_1.inject(browser_1.StatusBar),
        __metadata("design:type", Object)
    ], NotificationsContribution.prototype, "statusBar", void 0);
    NotificationsContribution = __decorate([
        inversify_1.injectable()
    ], NotificationsContribution);
    return NotificationsContribution;
}());
exports.NotificationsContribution = NotificationsContribution;
var NotificationsKeybindingContext = /** @class */ (function () {
    function NotificationsKeybindingContext() {
        this.id = NotificationsKeybindingContext_1.notificationsVisible;
    }
    NotificationsKeybindingContext_1 = NotificationsKeybindingContext;
    NotificationsKeybindingContext.prototype.isEnabled = function (_arg) {
        return this.manager.centerVisible || this.manager.toastsVisible;
    };
    var NotificationsKeybindingContext_1;
    __decorate([
        inversify_1.inject(notifications_manager_1.NotificationManager),
        __metadata("design:type", notifications_manager_1.NotificationManager)
    ], NotificationsKeybindingContext.prototype, "manager", void 0);
    NotificationsKeybindingContext = NotificationsKeybindingContext_1 = __decorate([
        inversify_1.injectable()
    ], NotificationsKeybindingContext);
    return NotificationsKeybindingContext;
}());
exports.NotificationsKeybindingContext = NotificationsKeybindingContext;
(function (NotificationsKeybindingContext) {
    NotificationsKeybindingContext.notificationsVisible = 'notificationsVisible';
})(NotificationsKeybindingContext = exports.NotificationsKeybindingContext || (exports.NotificationsKeybindingContext = {}));
exports.NotificationsKeybindingContext = NotificationsKeybindingContext;
