// @theia/task/lib/browser/task-definition-registry.spec.js
"use strict";
/********************************************************************************
 * Copyright (C) 2019 Ericsson and others.
 *
 * This program and the accompanying materials are made available under the
 * terms of the Eclipse Public License v. 2.0 which is available at
 * http://www.eclipse.org/legal/epl-2.0.
 *
 * This Source Code may also be made available under the following Secondary
 * Licenses when the conditions for such availability set forth in the Eclipse
 * Public License v. 2.0 are satisfied: GNU General Public License, version 2
 * with the GNU Classpath Exception which is available at
 * https://www.gnu.org/software/classpath/license.html.
 *
 * SPDX-License-Identifier: EPL-2.0 OR GPL-2.0 WITH Classpath-exception-2.0
 ********************************************************************************/
Object.defineProperty(exports, "__esModule", { value: true });
var chai_1 = require("chai");
var task_definition_registry_1 = require("./task-definition-registry");
// tslint:disable:no-unused-expression
describe('TaskDefinitionRegistry', function () {
    var registry;
    var definitonContributionA = {
        taskType: 'extA',
        source: 'extA',
        required: ['extensionType'],
        properties: {
            required: ['extensionType'],
            all: ['extensionType', 'taskLabel']
        }
    };
    var definitonContributionB = {
        taskType: 'extA',
        source: 'extA',
        properties: {
            required: ['extensionType', 'taskLabel', 'taskDetailedLabel'],
            all: ['extensionType', 'taskLabel', 'taskDetailedLabel']
        }
    };
    beforeEach(function () {
        registry = new task_definition_registry_1.TaskDefinitionRegistry();
    });
    describe('register function', function () {
        it('should transform the task definition contribution and store it in memory', function () {
            registry.register(definitonContributionA);
            chai_1.expect(registry['definitions'].get(definitonContributionA.taskType)).to.be.ok;
            chai_1.expect(registry['definitions'].get(definitonContributionA.taskType)[0]).to.deep.equal(definitonContributionA);
        });
    });
    describe('getDefinitions function', function () {
        it('should return all definitions associated with the given type', function () {
            registry.register(definitonContributionA);
            var defs1 = registry.getDefinitions(definitonContributionA.taskType);
            chai_1.expect(defs1.length).to.eq(1);
            registry.register(definitonContributionB);
            var defs2 = registry.getDefinitions(definitonContributionA.taskType);
            chai_1.expect(defs2.length).to.eq(2);
        });
    });
    describe('getDefinition function', function () {
        it('should return undefined if the given task configuration does not match any registered definitions', function () {
            registry.register(definitonContributionA);
            registry.register(definitonContributionB);
            var defs = registry.getDefinition({
                type: definitonContributionA.taskType, label: 'grunt task', task: 'build'
            });
            chai_1.expect(defs).to.be.not.ok;
        });
        it('should return the best match if there is one or more registered definitions match the given task configuration', function () {
            registry.register(definitonContributionA);
            registry.register(definitonContributionB);
            var defs = registry.getDefinition({
                type: definitonContributionA.taskType, label: 'extention task', extensionType: 'extensionType', taskLabel: 'taskLabel'
            });
            chai_1.expect(defs).to.be.ok;
            chai_1.expect(defs.taskType).to.be.eq(definitonContributionA.taskType);
            var defs2 = registry.getDefinition({
                type: definitonContributionA.taskType, label: 'extention task', extensionType: 'extensionType', taskLabel: 'taskLabel', taskDetailedLabel: 'taskDetailedLabel'
            });
            chai_1.expect(defs2).to.be.ok;
            chai_1.expect(defs2.taskType).to.be.eq(definitonContributionB.taskType);
        });
    });
});
