// sprotty-theia/lib/sprotty/theia-diagram-server.js
"use strict";
/********************************************************************************
 * Copyright (c) 2017-2018 TypeFox and others.
 *
 * This program and the accompanying materials are made available under the
 * terms of the Eclipse Public License v. 2.0 which is available at
 * http://www.eclipse.org/legal/epl-2.0.
 *
 * This Source Code may also be made available under the following Secondary
 * Licenses when the conditions for such availability set forth in the Eclipse
 * Public License v. 2.0 are satisfied: GNU General Public License, version 2
 * with the GNU Classpath Exception which is available at
 * https://www.gnu.org/software/classpath/license.html.
 *
 * SPDX-License-Identifier: EPL-2.0 OR GPL-2.0 WITH Classpath-exception-2.0
 ********************************************************************************/
var __extends = (this && this.__extends) || (function () {
    var extendStatics = Object.setPrototypeOf ||
        ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
        function (d, b) { for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p]; };
    return function (d, b) {
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
var __metadata = (this && this.__metadata) || function (k, v) {
    if (typeof Reflect === "object" && typeof Reflect.metadata === "function") return Reflect.metadata(k, v);
};
Object.defineProperty(exports, "__esModule", { value: true });
var inversify_1 = require("inversify");
var sprotty_1 = require("sprotty");
exports.IRootPopupModelProvider = Symbol('IRootPopupModelProvider');
/**
 * A sprotty DiagramServer that integrates with the Theia workbench.
 *
 * This class is the sprotty side of the Theia/sprotty integration. It
 * is instantiated with the DI container of the sprotty diagram. Theia
 * services are available via the TheiaDiagramServerConnector.
 */
var TheiaDiagramServer = /** @class */ (function (_super) {
    __extends(TheiaDiagramServer, _super);
    function TheiaDiagramServer() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    TheiaDiagramServer.prototype.connect = function (connector) {
        this._connector = connector;
    };
    TheiaDiagramServer.prototype.disconnect = function () {
    };
    Object.defineProperty(TheiaDiagramServer.prototype, "connector", {
        get: function () {
            if (!this._connector) {
                throw Error("TheiaDiagramServer is not connected.");
            }
            return this._connector;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(TheiaDiagramServer.prototype, "sourceUri", {
        get: function () {
            return this._sourceUri;
        },
        enumerable: true,
        configurable: true
    });
    TheiaDiagramServer.prototype.initialize = function (registry) {
        _super.prototype.initialize.call(this, registry);
        registry.register(sprotty_1.SelectAction.KIND, this);
    };
    TheiaDiagramServer.prototype.handle = function (action) {
        if (action instanceof sprotty_1.RequestModelAction && action.options !== undefined)
            this._sourceUri = action.options.sourceUri;
        return _super.prototype.handle.call(this, action);
    };
    TheiaDiagramServer.prototype.handleLocally = function (action) {
        if (action instanceof sprotty_1.RequestPopupModelAction) {
            return this.handleRequestPopupModel(action);
        }
        else {
            return _super.prototype.handleLocally.call(this, action);
        }
    };
    TheiaDiagramServer.prototype.handleExportSvgAction = function (action) {
        this.connector.save(this.sourceUri, action);
        return true;
    };
    TheiaDiagramServer.prototype.handleRequestPopupModel = function (action) {
        var _this = this;
        if (action.elementId === this.currentRoot.id) {
            this.rootPopupModelProvider.getPopupModel(action, this.currentRoot).then(function (model) {
                if (model)
                    _this.actionDispatcher.dispatch(new sprotty_1.SetPopupModelAction(model));
            });
            return false;
        }
        else {
            return true;
        }
    };
    TheiaDiagramServer.prototype.handleServerStateAction = function (status) {
        this.connector.showStatus(this.clientId, status);
        return false;
    };
    TheiaDiagramServer.prototype.sendMessage = function (message) {
        this.connector.sendMessage(message);
    };
    /**
     * made public
     */
    TheiaDiagramServer.prototype.messageReceived = function (message) {
        _super.prototype.messageReceived.call(this, message);
    };
    __decorate([
        inversify_1.inject(exports.IRootPopupModelProvider), inversify_1.optional(),
        __metadata("design:type", Object)
    ], TheiaDiagramServer.prototype, "rootPopupModelProvider", void 0);
    TheiaDiagramServer = __decorate([
        inversify_1.injectable()
    ], TheiaDiagramServer);
    return TheiaDiagramServer;
}(sprotty_1.DiagramServer));
exports.TheiaDiagramServer = TheiaDiagramServer;
