// sprotty-theia/lib/sprotty/theia-sprotty-selection-forwarder.js
"use strict";
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
var __metadata = (this && this.__metadata) || function (k, v) {
    if (typeof Reflect === "object" && typeof Reflect.metadata === "function") return Reflect.metadata(k, v);
};
Object.defineProperty(exports, "__esModule", { value: true });
/********************************************************************************
 * Copyright (c) 2018 EclipseSource and others.
 *
 * This program and the accompanying materials are made available under the
 * terms of the Eclipse Public License v. 2.0 which is available at
 * http://www.eclipse.org/legal/epl-2.0.
 *
 * This Source Code may also be made available under the following Secondary
 * Licenses when the conditions for such availability set forth in the Eclipse
 * Public License v. 2.0 are satisfied: GNU General Public License, version 2
 * with the GNU Classpath Exception which is available at
 * https://www.gnu.org/software/classpath/license.html.
 *
 * SPDX-License-Identifier: EPL-2.0 OR GPL-2.0 WITH Classpath-exception-2.0
 ********************************************************************************/
var inversify_1 = require("inversify");
var sprotty_1 = require("sprotty");
var core_1 = require("@theia/core");
function isSprottySelection(object) {
    return object !== undefined && object.selectedElementsIDs !== undefined
        && object.widgetId !== undefined;
}
exports.isSprottySelection = isSprottySelection;
var TheiaSprottySelectionForwarder = /** @class */ (function () {
    function TheiaSprottySelectionForwarder() {
    }
    TheiaSprottySelectionForwarder.prototype.initialize = function (registry) {
        registry.register(sprotty_1.RequestModelAction.KIND, this);
        registry.register(sprotty_1.SelectAction.KIND, this);
    };
    TheiaSprottySelectionForwarder.prototype.handle = function (action) {
        if (action instanceof sprotty_1.SelectAction) {
            this.selectionService.selection = {
                selectedElementsIDs: action.selectedElementsIDs,
                widgetId: this.viewerOptions.baseDiv,
                sourceUri: this.sourceUri
            };
        }
        else if (action instanceof sprotty_1.RequestModelAction && action.options !== undefined) {
            this.sourceUri = action.options.sourceUri;
        }
    };
    __decorate([
        inversify_1.inject(sprotty_1.TYPES.ViewerOptions),
        __metadata("design:type", Object)
    ], TheiaSprottySelectionForwarder.prototype, "viewerOptions", void 0);
    __decorate([
        inversify_1.inject(core_1.SelectionService),
        __metadata("design:type", core_1.SelectionService)
    ], TheiaSprottySelectionForwarder.prototype, "selectionService", void 0);
    TheiaSprottySelectionForwarder = __decorate([
        inversify_1.injectable()
    ], TheiaSprottySelectionForwarder);
    return TheiaSprottySelectionForwarder;
}());
exports.TheiaSprottySelectionForwarder = TheiaSprottySelectionForwarder;
