// @typefox/monaco-editor-core/esm/vs/base/browser/ui/tree/tree.js
/*---------------------------------------------------------------------------------------------
 *  Copyright (c) Microsoft Corporation. All rights reserved.
 *  Licensed under the MIT License. See License.txt in the project root for license information.
 *--------------------------------------------------------------------------------------------*/
export var TreeMouseEventTarget;
(function (TreeMouseEventTarget) {
    TreeMouseEventTarget[TreeMouseEventTarget["Unknown"] = 0] = "Unknown";
    TreeMouseEventTarget[TreeMouseEventTarget["Twistie"] = 1] = "Twistie";
    TreeMouseEventTarget[TreeMouseEventTarget["Element"] = 2] = "Element";
})(TreeMouseEventTarget || (TreeMouseEventTarget = {}));
export var TreeDragOverReactions = {
    acceptBubbleUp: function () { return { accept: true, bubble: 1 /* Up */ }; },
    acceptBubbleDown: function (autoExpand) {
        if (autoExpand === void 0) { autoExpand = false; }
        return { accept: true, bubble: 0 /* Down */, autoExpand: autoExpand };
    },
    acceptCopyBubbleUp: function () { return { accept: true, bubble: 1 /* Up */, effect: 0 /* Copy */ }; },
    acceptCopyBubbleDown: function (autoExpand) {
        if (autoExpand === void 0) { autoExpand = false; }
        return { accept: true, bubble: 0 /* Down */, effect: 0 /* Copy */, autoExpand: autoExpand };
    }
};
